base/abstractarray.jl

```diff
@@ -422,6 +422,21 @@ function _maxlength(A, B, C...)
     max(length(A), _maxlength(B, C...))
 end
 
+# eachindex(A, d) and eachindex(A, (:, j))
+eachindex(A::AbstractArray, d::Integer) = 1:size(A, d)
```

Comment:
It shouldn't conflict because `B` is an `AbstractArray` (or should be; if we haven't specified it, we probably should). And it's nearly certain that we'll be deprecating that interface (though obviously we can't just discard it).

The way I'm thinking of it, `eachindex` is becoming the companion of `getindex` and `setindex!`. Those functions are much, much more overloaded than `eachindex` is right now. This PR takes `eachindex` more in the direction of `getindex`, and if we deprecate `eachindex(A, B)` the similarities will be easier to explore.

I don't think keyword arguments work with iterators: they are invariably performance-critical.